mason new --moduleName

ankingcodes opened the pull request:
Added a flag `--moduleName` to `mason new` and `mason init` to allow illegal package names in Mason.

Added support for using the command as follows : 
```
mason new illegal-name --moduleName=legalName
# OR
mason new illegal-name --moduleName legalName
```
The same goes for `mason init`. 
Added relevant tests, updated `Mason Help` and added `docs`

fixes #14955

ankingcodes commented:
@ben-albrecht This is ready for review 👍

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -35,8 +35,12 @@ proc masonNew(args) throws {
       var name = '';
+      var i = 0;
       for arg in args[2..] {
+        i = i + 1;
```
```suggestion
        i += 1;
```

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -35,8 +35,12 @@ proc masonNew(args) throws {
+      var isModuleName = false;
+      var moduleName = '';
       var name = '';
+      var i = 0;
```
Would it make sense to start this counter at 2 so we don't have to offset elsewhere?

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -47,18 +51,40 @@ proc masonNew(args) throws {
+          else {
+            moduleName = args[i+2];
+            name = args[2];
+            break;
```
If we are breaking, does that end arg parsing? What about arguments that come after `--moduleName foo`?

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -47,18 +51,40 @@ proc masonNew(args) throws {
+          if isDir(name) {
+            throw new owned MasonError("A directory named '" + name + "' already exists");
+          }
+          InitProject(name, moduleName='', isModuleName, vcs, show);
```
It may be cleaner to always define `moduleName` and `name` regardless of whether `--moduleName` was passed.

Then we could reduce some branching in the code and call `InitProject` with both args:

```
InitProject(name, moduleName, vcs, show);
```

Also, I think `packageName` would be a more accurate identifier, since it impacts both the Chapel module name and the `[brick][name]` in `Mason.toml`. We can also consider renaming `name` to `dirName` to reduce ambiguity.

ben-albrecht replied:
This could also eliminate the need for `isModuleName` entirely. We would know if `--moduleName` was called by checking if `moduleName == name` (or `packageName == dirName` with my suggested renames).

ankingcodes replied:
> Also, I think packageName would be a more accurate identifier, since it impacts both the Chapel module name and the [brick][name] in Mason.toml

@ben-albrecht The `--moduleName` flag should only change the `.chpl` file inside of `src/` right ? I'm unclear about what  it has to do with `[brick][name]`? Shouldn't brick-name be the same as that of the project's `cwd` ? 
For example , `mason new project-testSrc --moduleName=project` should perform as follows: 

```
project-testSrc/
    Mason.toml      # [brick][name] = "project-testSrc"
    src/
         project.chpl         # module project { ...... }
    example/
    test/
```

ben-albrecht replied:
I believe we want the brick name to match the Chapel module name

ankingcodes replied:
> This could also eliminate the need for isModuleName entirely

@ben-albrecht  Even if we get rid of `isModuleName`, we will require it to avoid overwriting of name in the last condition as : 
```
...
 else if arg.startsWith('--moduleName') {
   isModuleName = true;
   if arg.startsWith('--moduleName=') {
     var res = arg.split("=");
     moduleName = res[2];
     name = args[2];
   }
   else {
    moduleName = args[countArgs];
    name = args[2];
   }
 }
 else {
 # Name got overwritten , need to use isModuleName check here
  name = arg;
}
...
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -227,3 +268,20 @@ proc addSection(sectionName: string, path: string, tomlFile: unmanaged Toml, sho
   if show then writeln("Added [" + sectionName + "] section to Mason.toml");
 }
+
+proc makeLegalModule(path: string, moduleName: string) throws {
```
Perhaps this should just be `createModule`, reiterating the `--moduleName` can be used for other purposes other than using illegal directory names.

Also, can you add a doc-string to this function?

ankingcodes commented:
@ben-albrecht I had encountered a case while testing the command. 
```
mkdir proj-src
mason init proj-src --name proj  
# get into project directory file and remove something
cd proj-src && rm -rf src Mason.toml    
mason init        # generates error   
``` 
Earlier, in the `mason init` command, we would first get the current working directory, validate its name for legal/illegal, then proceed with creating files without overwriting. 
Now, `--name` flag allows us to create project-directory with an illegal name.  
During developement of `--name` we had added a check to give error warning if a Mason.toml file is already present. However, there can be case when Mason.toml file is missing. Then, using `mason init` gives an error for `Bad package name`. 

Also , 
```
mkdir proj-src
mason init proj-src --name proj  
# get into project directory file and remove something
cd proj-src && rm -rf example  
mason init        # generates error   
``` 
Even this generates a `Bad package name` error, since `--name` is not used and illegal current working directory  name is being validated. 

I believe we have two ways to solve this : 
- Improve the error message, asking user to check for legal name in Mason.toml and use the flag.
- Improve the validation function, to check for legal file name in Mason.toml when being called.

Which one would you recommend ?

ben-albrecht commented:
@ankingcodes - I think we should determine the project name using the following precedence:

1. `--name` (if thrown)
2. `Mason.toml` `[brick][name]` value (if found)
3. File name of `src/<project>.chpl` (if found)
4. Top-level directory name (always available)

I suggest we define the project name based on that precedence, then validate if it is a legal name, emitting an error if it is not. Of course, we should continue to emit an error if there are incompatibilities between project name, Chapel module name, and/or `--name` value.

ankingcodes commented:
@ben-albrecht thanks for this very comprehensive comment ! 
Also, would you encourage separate functions for `init` and `new` ? because error messages should be different. For example, 
if an user types `mason init proj-src` he gets an error as follows : 
```
Bad package name 'proj-src' - only Chapel identifiers are legal package names.
Please use mason new <illegal-name> --name <legal-name>
```
The error mentions `mason new`.

ben-albrecht commented:
> Also, would you encourage separate functions for init and new ? because error messages should be different. For example

ankingcodes commented:
> It would be nice to not implement this logic twice.

I don't think this logic is required for `mason new`, since it doesn't have many use cases. We require a separate check only for `mason init`.

ben-albrecht commented:
> @ben-albrecht I don't think this logic is required for mason new, since it doesn't have many use cases. We require a separate check only for mason init.

Good point!

ankingcodes commented:
@ben-albrecht modified & tested `mason init`. I think it's ready for a second round of review. 

Also, some tests (c1cddfd) involving `mason new` failed on my machine. However the command works perfectly. I used a hack to succeed those tests. Please check those in particular. 
Thanks !

ben-albrecht commented:
@ankingcodes - I am looking over the code and trying it out. In the meantime, here are some failures I saw:

```
> start_test test/mason
[Test Summary - 200302.121647]
[Error matching program output for test/mason/masonNewBadName (execopts: 2)]
[Error matching program output for test/mason/masonNewBadName (execopts: 3)]
[Error matching program output for test/mason/masonNewBadName (execopts: 4)]
[Error matching program output for test/mason/masonNewBadName (execopts: 5)]
[Error matching program output for test/mason/masonNewBadName (execopts: 6)]
[Error matching program output for test/mason/masonNewTest]
[Error matching program output for test/mason/chplVersion/mason-chpl-version-new]
[Error matching program output for test/mason/publish-tests/badDry]
[Error matching program output for test/mason/publish-tests/badRegTest]
[Error matching program output for test/mason/publish-tests/dryTest]
[Error matching program output for test/mason/publish-tests/publish]
[Error matching program output for test/mason/publish-tests/publishGitCheck]
[Summary: #Successes = 70 | #Failures = 12 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]
```

ankingcodes commented:
@ben-albrecht fixed tests 👍

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -67,64 +92,66 @@ proc masonNew(args) throws {
 }
 
-proc InitProject(name, vcs, show) throws {
+proc InitProject(dirName, packageName, vcs, show) throws {
```
I know you didn't introduce these functions, but bonus points if you add some short doc-strings to them (following the mantra of "cleaning up code you modify").

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -67,64 +92,66 @@ proc masonNew(args) throws {
-    makeExampleDir(name);  
-    writeln("Created new library project: " + name);
+  if isDir(dirName) {
+    if dirName!=packageName then makeBasicToml(packageName, path=dirName);
```
We shouldn't need to check if they are equal. If they are then, the calls are the same:

```suggestion
   makeBasicToml(packageName, path=dirName);
```

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -67,64 +92,66 @@ proc masonNew(args) throws {
+    if dirName!=packageName then makeBasicToml(packageName, path=dirName);
+    else makeBasicToml(dirName, path=dirName);
+    makeSrcDir(dirName);
+    if dirName!=packageName then makeModule(dirName, fileName=packageName);
```
We shouldn't need to check if they are equal. If they are then, the calls are the same:

```suggestion
   makeModule(dirName, fileName=packageName);
```

ben-albrecht commented on tools/mason/MasonNew.chpl:
```diff
@@ -47,18 +50,40 @@ proc masonNew(args) throws {
+          }
+        }
+      }
+      if dirName != packageName then {
```
We shouldn't need to check if they are equal. If they are then, the calls are the same.

ben-albrecht replied:
Let's add a newline here to visually separate this from the arg parsing

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,43 +37,90 @@ Initialises a library project in a project directory
-          validateMasonFile(path, basename(path), show);
-          var isInitialized = validateInit(path, show);
-          if isInitialized > 0 then
+        if dirName != packageName then {
```
We shouldn't need to branch here. The calls should be the same if regardless of whether `dirName == packageName` or not.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,43 +37,90 @@ Initialises a library project in a project directory
-        var isInitialized = validateInit(path, show);
-        if isInitialized > 0 then
-          writeln("Initialized new library project: " + name);
+      if dirName != packageName then {
```
We shouldn't need to branch here. The calls should be the same if regardless of whether `dirName == packageName` or not.

ankingcodes commented:
@ben-albrecht  Fixed changes requested:
- fixed arg-parsing bug on tests
- refactored `if-else` statements of `mason new` and `mason init` to `select`
- shortened code and avoided differentiation on the basis of `packageName` and `dirName`

ben-albrecht commented on test/mason/masonInit/masonInitPathTest.chpl:
```diff
@@ -5,9 +5,10 @@ use MasonInit;
 
 proc main(){
-  const args = ['init','testSrc'];
+  const args: [0..2] string;
```
Ah, so the issue was that `masonInit` expects args to be 0-based. Would it be difficult to instead make `masonInit` support any indices? You can get the low value with `args.domain.low` for example.

ankingcodes replied:
@ben-albrecht I applied `args.domain.low` to the code. Now, it works fine for `mason new projectName` or `mason init projectName` but doesn't work for `mason init project-Name --name project` or for mason new. However, the typed command on terminal works perfectly. I think we keep the `args` declarations for tests involving the flag.

ankingcodes replied:
In `mason new project-Name --name project` , it creates a module by name of new.

ben-albrecht replied:
>  but doesn't work for `mason init project-Name --name project` or for `mason new`.

That seems odd. Is that just a bug in the arg parser then?

ben-albrecht commented on test/mason/masonInit/masonInitTest4.chpl:
```diff
@@ -4,10 +4,12 @@ use MasonUtils;
+  initArgs = ['mason','init','testSrc'];
   masonInit(initArgs);
   
   runCommand('rm -rf testSrc');
```
Mind updating the remaining `masonInit` tests to use `rmTree` over `runCommand`?

ben-albrecht commented on test/mason/masonInit/masonInitTest3.chpl:
```diff
@@ -4,15 +4,17 @@ use MasonUtils;
+  const initArgs : [0..2] string;
+  newArgs = ['mason','new','testSrc'];
   masonNew(newArgs);
   runCommand('rm -rf testSrc/.git');
```
```suggestion
  rmTree(testSrc/.git');
```

ben-albrecht commented on test/mason/masonNewModule.chpl:
```diff
@@ -0,0 +1,14 @@
+  if isFile("./project-testSrc/src/project.chpl") {
+    writeln("Project.chpl has been successfully created");
+  }
+  runCommand('rm -rf project-testSrc');
```
```suggestion
  rmTree('project-testSrc');
```

ben-albrecht commented on test/mason/publish-tests/badDry.chpl:
```diff
@@ -6,7 +6,7 @@ use MasonNew;
 
 proc main() {
-  masonNew(['mason', 'new', 'publishCheck']);
+  masonNew(['new', 'publishCheck','']);
```
I think we can revert these back to how they were now, unless I misunderstand the reason for changing these.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,71 @@ Initialises a library project in a project directory
+    var dirName = '';
     var show = false;
+    var packageName = '';
+    var countArgs = 2;
```
I think this should work?

```suggestion
    var countArgs = args.domain.low + 2;
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,71 @@ Initialises a library project in a project directory
     var show = false;
+    var packageName = '';
+    var countArgs = 2;
     for arg in args[2..] {
```
```suggestion
    for arg in args[args.domain.low+2..] {
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,71 @@ Initialises a library project in a project directory
+            packageName = res[2];
+            dirName = args[2];
+          }
+          if packageName.length == 0 then {
```
length is being deprecated soon:

```suggestion
          if packageName.size == 0 then {
```

ben-albrecht replied:
This check depends on previous flags, which results in a flag order dependence, i.e. will these produce the same result?

```
mason init --name=foo bar
mason init bar --name=foo
```

I would suggest having `--name` set only the `packageName`, and this arg set only the `dirName`. After arg parsing is complete, you can set the `dirName` if it is not set, and then set the `packageName` it is not set.

ben-albrecht replied:
Same goes for `MasonNew`

ankingcodes replied:
>> mason init --name=foo bar

I don't think an user would actually use the flag like this.
A new user would typically use `mason init illegal-name`. Then, after seeing the error message, he would follow correct syntax `mason init illegal-name --name legalName`.
I don't think an experienced user would use the flag before, unless deliberately. 
If this is not absolutely important, let's keep it as is.

ben-albrecht replied:
I agree that it's an odd usage, but I think a command line tool's arg parser should be robust enough to handle the arguments in any order (beyond subcommands, which require a specific order, e.g. `mason init` and not `mason --flag init`. 

I am not sure what the expected behavior of `mason init --name=foo bar` would be otherwise.

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,71 @@ Initialises a library project in a project directory
+          dirName = args[2];
+        }
+        otherwise {
+          if arg.startsWith('--name') {
```
You could make this clearer by specifying the exact case we are checking for here (since a previous case checks for `--name`:

```suggestion
          if arg.startsWith('--name=') {
```

ben-albrecht replied:
Same goes for `MasonNew`

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -102,50 +140,50 @@ proc validateInit(path: string, show: bool) throws {
+    }
     else {
-        if isDir(path + dir) == false {
+      if isDir(path + dir) == false {
```
```suggestion
      if !isDir(path + dir) {
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -102,50 +140,50 @@ proc validateInit(path: string, show: bool) throws {
       }
     }
+    else if dir == "/src/" + moduleName {
+      if isFile(path + dir) == false {
```
```suggestion
      if !isFile(path + dir) {
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,71 @@ Initialises a library project in a project directory
     }
-    if name == '' {
+
+    if dirName.startsWith('--name') then dirName = '';
```
Does this ever actually happen? I would expect it not to.

ankingcodes replied:
Without this condition, 
```
mkdir project-Src
cd project-Src
mason init --name=project      # packageName = project , dirName = --name=project
ERROR !
```
With this condition, we use packageName to initialise the module

ben-albrecht replied:
shouldn't  the case above catch this and set `dir` appropriately though? 

` if arg.startsWith('--name=') { ... }`

https://github.com/chapel-lang/chapel/pull/14997/files#diff-ef69f825f0dea9e83ad07d0e0f9ae4acR63

ankingcodes commented:
@ben-albrecht just a ping to check the comments and previous commits. This has been lying around for quite some time.

ben-albrecht commented:
Thanks @ankingcodes - I am planning on taking a look at this today.

ankingcodes commented:
@ben-albrecht resolved arg parsing issues. 
All tests related to the PR passed other than 1, related to `mason-doc`. Please check if you get the same. Other than that, I believe this is ready to merge.
```
[Test Summary - 200314.205044] 
[Error matching program output for mason/mason-doc/docPkg]
[Summary: #Successes = 81 | #Failures = 1 | #Futures = 0 | #Warnings = 0 ]
[Summary: #Passing Suppressions = 0 | #Passing Futures = 0 ]
[END]
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,66 @@ Initialises a library project in a project directory
-        if isInitialized > 0 then
-          writeln("Initialized new library project: " + name);
-      }
+      if packageName.length > 0 then name = packageName;
```
```suggestion
      if packageName.size > 0 then name = packageName;
```

ben-albrecht commented on tools/mason/MasonInit.chpl:
```diff
@@ -37,44 +37,66 @@ Initialises a library project in a project directory
-            writeln("Initialized new library project in " + path + ": " + basename(path));
-        }
+        var name = basename(dirName);
+        if packageName.length > 0 then name = packageName;
```
```suggestion
        if packageName.size > 0 then name = packageName;
```

ankingcodes commented:
@ben-albrecht made changes 👍

ben-albrecht commented:
>  made changes 👍

Thanks! Double checking testing before approving/merging.
